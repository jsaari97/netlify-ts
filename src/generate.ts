import { pullCollection } from "./collection";
import { appendExport, formatType } from "./output";
import { resolveRelations, resolveWidget, transformType } from "./widget";
import type { NetlifyCMSConfig, NetlifyTsOptions } from "./types";

export const generateTypes = (config: NetlifyCMSConfig, options: NetlifyTsOptions = {}): string => {
  const externalMediaLibrary = hasExternalMediaLibrary(config);

  return config.collections
    .flatMap(pullCollection)
    .map(resolveWidget({ externalMediaLibrary }))
    .reduce(
      transformType({
        label: !!options.label,
        capitalize: !!options.capitalize,
        delimiter: options.delimiter,
      }),
      [[], []],
    )
    .flat()
    .map(resolveRelations({ delimiter: options.delimiter }))
    .map(formatType)
    .map(appendExport)
    .join("\n\n")
    .replace(/^/, "/* eslint-disable */\n/* tslint:disable */\n\n")
    .concat("\n");
};

export const hasExternalMediaLibrary = (config: NetlifyCMSConfig): boolean =>
  !!config.media_library?.name && !!config.media_library?.config;
