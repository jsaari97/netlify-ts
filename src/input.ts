import yaml from "js-yaml";
import fs from "fs";
import path from "path";
import type { Collection, NetlifyCMSConfig } from "./types";

export const loadConfig = (config: string | NetlifyCMSConfig): Collection[] => {
  let data;

  if (typeof config === "object") {
    data = config;
  }

  if (typeof config === "string") {
    data = loadConfigFile(config);
  }

  if (typeof data !== "object" || !data.collections) {
    throw new Error("Failed loading collections from config");
  }

  return data.collections;
};

const loadConfigFile = (fileName: string): NetlifyCMSConfig => {
  if (!path.extname(fileName).match(/\.ya?ml$/)) {
    throw new Error("Invalid filetype, must be an yaml file");
  }

  const file = fs.readFileSync(path.resolve(process.cwd(), fileName), "utf8");

  return yaml.load(file) as NetlifyCMSConfig;
};
