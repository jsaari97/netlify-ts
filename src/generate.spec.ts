import { loadConfig } from "./input";
import { generateTypes } from "./generate";

describe("Output testing", () => {
  it("should parse kitchen sink correctly", () => {
    const config = loadConfig("kitchen-sink.yml");

    expect(generateTypes(config, { label: false })).toMatchSnapshot();
  });

  it("should parse kitchen sink with label option", () => {
    const config = loadConfig("kitchen-sink.yml");

    expect(generateTypes(config, { label: true })).toMatchSnapshot();
  });

  it("should support capitalization of type names", () => {
    const config = loadConfig("kitchen-sink.yml");

    expect(generateTypes(config, { capitalize: true })).toMatchSnapshot();
  });

  it("should support custom delimiter for type names", () => {
    const config = loadConfig("kitchen-sink.yml");

    expect(generateTypes(config, { delimiter: "-" })).toMatchSnapshot();
  });

  it("should support label, capitalization and delimiter at the same time", () => {
    const config = loadConfig("kitchen-sink.yml");

    expect(
      generateTypes(config, { label: true, capitalize: true, delimiter: "" }),
    ).toMatchSnapshot();
  });
});
