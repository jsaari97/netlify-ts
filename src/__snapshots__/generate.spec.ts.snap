// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output testing should parse kitchen sink correctly 1`] = `
"/* eslint-disable */
/* tslint:disable */

export interface posts {
  title: string;
  draft: boolean;
  date: string;
  image?: string;
}

export interface faq {
  title: string;
}

export interface general_posts {
  front_limit: string;
  author: string;
  thumb?: string;
}

export interface general {
  site_title: string;
  posts: general_posts;
}

export interface authors_authors {
  name: string;
  description: string;
}

export interface authors {
  authors: authors_authors[];
}

export type kitchenSink_select_options = "a" | "b" | "c";

export type kitchenSink_select_multiple_options = "a" | "b" | "c";

export type kitchenSink_select_numeric_options = 1 | 2 | 3;

export type kitchenSink_object_select_options = "a" | "b" | "c";

export interface kitchenSink_object {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink_object_select_options;
}

export type kitchenSink_list_select_options = "a" | "b" | "c";

export type kitchenSink_list_object_select_options = "a" | "b" | "c";

export type kitchenSink_list_object_list_select_options = "a" | "b" | "c";

export type kitchenSink_list_object_list_object_select_options = "a" | "b" | "c";

export interface kitchenSink_list_object_list_object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink_list_object_list_object_select_options;
}

export interface kitchenSink_list_object_list {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink_list_object_list_select_options;
  hidden: any;
  object: kitchenSink_list_object_list_object;
}

export interface kitchenSink_list_object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink_list_object_select_options;
  list: kitchenSink_list_object_list[];
}

export interface kitchenSink_list {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink_list_select_options;
  object: kitchenSink_list_object;
}

export interface kitchenSink_typed_list_type_2_object_nested {
  number: string;
}

export type kitchenSink_typed_list_type_2_object_select_options = "a" | "b" | "c";

export interface kitchenSink_typed_list_type_1_object {
  type: "type_1_object";
  string: string;
  boolean: boolean;
  text: string;
}

export interface kitchenSink_typed_list_type_2_object {
  type: "type_2_object";
  number: string;
  nested: kitchenSink_typed_list_type_2_object_nested;
  select: kitchenSink_typed_list_type_2_object_select_options;
  datetime: string;
  markdown: string;
}

export interface kitchenSink_typed_list_type_3_object {
  type: "type_3_object";
  date: string;
  image: string;
  file: string;
}

export interface kitchenSink_typed_list_type_4_object {
  type: "type_4_object";
  name: string;
}

export interface kitchenSink_code {
  code: string;
  lang: string;
}

export interface kitchenSink_code_alt {
  alt: string;
  typescript: string;
}

export interface kitchenSink {
  post: any;
  title: string;
  boolean: boolean;
  map: string;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  color: string;
  colorEditable: string;
  image: string;
  file: string;
  select: kitchenSink_select_options;
  select_multiple: kitchenSink_select_multiple_options[];
  select_numeric: kitchenSink_select_numeric_options;
  hidden: any;
  object: kitchenSink_object;
  list: kitchenSink_list[];
  typed_list: (kitchenSink_typed_list_type_1_object | kitchenSink_typed_list_type_2_object | kitchenSink_typed_list_type_3_object | kitchenSink_typed_list_type_4_object)[];
  code: kitchenSink_code;
  code_alt: kitchenSink_code_alt;
  snippet: string;
}
"
`;

exports[`Output testing should parse kitchen sink with label option 1`] = `
"/* eslint-disable */
/* tslint:disable */

export interface Post {
  title: string;
  draft: boolean;
  date: string;
  image?: string;
}

export interface FAQ {
  title: string;
}

export interface SiteSettings_PostSettings {
  front_limit: string;
  author: string;
  thumb?: string;
}

export interface SiteSettings {
  site_title: string;
  posts: SiteSettings_PostSettings;
}

export interface Authors_Author {
  name: string;
  description: string;
}

export interface Authors {
  authors: Authors_Author[];
}

export type KitchenSink_Select_options = "a" | "b" | "c";

export type KitchenSink_SelectMultiple_options = "a" | "b" | "c";

export type KitchenSink_SelectNumeric_options = 1 | 2 | 3;

export type KitchenSink_Object_Select_options = "a" | "b" | "c";

export interface KitchenSink_Object {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_Object_Select_options;
}

export type KitchenSink_List_Select_options = "a" | "b" | "c";

export type KitchenSink_List_Object_Select_options = "a" | "b" | "c";

export type KitchenSink_List_Object_List_Select_options = "a" | "b" | "c";

export type KitchenSink_List_Object_List_Object_Select_options = "a" | "b" | "c";

export interface KitchenSink_List_Object_List_Object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Object_List_Object_Select_options;
}

export interface KitchenSink_List_Object_List {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Object_List_Select_options;
  hidden: any;
  object: KitchenSink_List_Object_List_Object;
}

export interface KitchenSink_List_Object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Object_Select_options;
  list: KitchenSink_List_Object_List[];
}

export interface KitchenSink_List {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Select_options;
  object: KitchenSink_List_Object;
}

export interface KitchenSink_TypedList_Type2Object_NestedObject {
  number: string;
}

export type KitchenSink_TypedList_Type2Object_Select_options = "a" | "b" | "c";

export interface KitchenSink_TypedList_Type1Object {
  type: "type_1_object";
  string: string;
  boolean: boolean;
  text: string;
}

export interface KitchenSink_TypedList_Type2Object {
  type: "type_2_object";
  number: string;
  nested: KitchenSink_TypedList_Type2Object_NestedObject;
  select: KitchenSink_TypedList_Type2Object_Select_options;
  datetime: string;
  markdown: string;
}

export interface KitchenSink_TypedList_type3Object {
  type: "type_3_object";
  date: string;
  image: string;
  file: string;
}

export interface KitchenSink_TypedList_type4Object {
  type: "type_4_object";
  name: string;
}

export interface KitchenSink_CodeBlock {
  code: string;
  lang: string;
}

export interface KitchenSink_CodeAltBlock {
  alt: string;
  typescript: string;
}

export interface KitchenSink {
  post: any;
  title: string;
  boolean: boolean;
  map: string;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  color: string;
  colorEditable: string;
  image: string;
  file: string;
  select: KitchenSink_Select_options;
  select_multiple: KitchenSink_SelectMultiple_options[];
  select_numeric: KitchenSink_SelectNumeric_options;
  hidden: any;
  object: KitchenSink_Object;
  list: KitchenSink_List[];
  typed_list: (KitchenSink_TypedList_Type1Object | KitchenSink_TypedList_Type2Object | KitchenSink_TypedList_type3Object | KitchenSink_TypedList_type4Object)[];
  code: KitchenSink_CodeBlock;
  code_alt: KitchenSink_CodeAltBlock;
  snippet: string;
}
"
`;

exports[`Output testing should support capitalization of type names 1`] = `
"/* eslint-disable */
/* tslint:disable */

export interface Posts {
  title: string;
  draft: boolean;
  date: string;
  image?: string;
}

export interface Faq {
  title: string;
}

export interface General_Posts {
  front_limit: string;
  author: string;
  thumb?: string;
}

export interface General {
  site_title: string;
  posts: General_Posts;
}

export interface Authors_Authors {
  name: string;
  description: string;
}

export interface Authors {
  authors: Authors_Authors[];
}

export type KitchenSink_Select_Options = "a" | "b" | "c";

export type KitchenSink_Select_Multiple_Options = "a" | "b" | "c";

export type KitchenSink_Select_Numeric_Options = 1 | 2 | 3;

export type KitchenSink_Object_Select_Options = "a" | "b" | "c";

export interface KitchenSink_Object {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_Object_Select_Options;
}

export type KitchenSink_List_Select_Options = "a" | "b" | "c";

export type KitchenSink_List_Object_Select_Options = "a" | "b" | "c";

export type KitchenSink_List_Object_List_Select_Options = "a" | "b" | "c";

export type KitchenSink_List_Object_List_Object_Select_Options = "a" | "b" | "c";

export interface KitchenSink_List_Object_List_Object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Object_List_Object_Select_Options;
}

export interface KitchenSink_List_Object_List {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Object_List_Select_Options;
  hidden: any;
  object: KitchenSink_List_Object_List_Object;
}

export interface KitchenSink_List_Object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Object_Select_Options;
  list: KitchenSink_List_Object_List[];
}

export interface KitchenSink_List {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSink_List_Select_Options;
  object: KitchenSink_List_Object;
}

export interface KitchenSink_Typed_List_Type_2_Object_Nested {
  number: string;
}

export type KitchenSink_Typed_List_Type_2_Object_Select_Options = "a" | "b" | "c";

export interface KitchenSink_Typed_List_Type_1_Object {
  type: "type_1_object";
  string: string;
  boolean: boolean;
  text: string;
}

export interface KitchenSink_Typed_List_Type_2_Object {
  type: "type_2_object";
  number: string;
  nested: KitchenSink_Typed_List_Type_2_Object_Nested;
  select: KitchenSink_Typed_List_Type_2_Object_Select_Options;
  datetime: string;
  markdown: string;
}

export interface KitchenSink_Typed_List_Type_3_Object {
  type: "type_3_object";
  date: string;
  image: string;
  file: string;
}

export interface KitchenSink_Typed_List_Type_4_Object {
  type: "type_4_object";
  name: string;
}

export interface KitchenSink_Code {
  code: string;
  lang: string;
}

export interface KitchenSink_Code_Alt {
  alt: string;
  typescript: string;
}

export interface KitchenSink {
  post: any;
  title: string;
  boolean: boolean;
  map: string;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  color: string;
  colorEditable: string;
  image: string;
  file: string;
  select: KitchenSink_Select_Options;
  select_multiple: KitchenSink_Select_Multiple_Options[];
  select_numeric: KitchenSink_Select_Numeric_Options;
  hidden: any;
  object: KitchenSink_Object;
  list: KitchenSink_List[];
  typed_list: (KitchenSink_Typed_List_Type_1_Object | KitchenSink_Typed_List_Type_2_Object | KitchenSink_Typed_List_Type_3_Object | KitchenSink_Typed_List_Type_4_Object)[];
  code: KitchenSink_Code;
  code_alt: KitchenSink_Code_Alt;
  snippet: string;
}
"
`;

exports[`Output testing should support custom delimiter for type names 1`] = `
"/* eslint-disable */
/* tslint:disable */

export interface posts {
  title: string;
  draft: boolean;
  date: string;
  image?: string;
}

export interface faq {
  title: string;
}

export interface general-posts {
  front_limit: string;
  author: string;
  thumb?: string;
}

export interface general {
  site_title: string;
  posts: general-posts;
}

export interface authors-authors {
  name: string;
  description: string;
}

export interface authors {
  authors: authors-authors[];
}

export type kitchenSink-select-options = "a" | "b" | "c";

export type kitchenSink-select-multiple-options = "a" | "b" | "c";

export type kitchenSink-select-numeric-options = 1 | 2 | 3;

export type kitchenSink-object-select-options = "a" | "b" | "c";

export interface kitchenSink-object {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink-object-select-options;
}

export type kitchenSink-list-select-options = "a" | "b" | "c";

export type kitchenSink-list-object-select-options = "a" | "b" | "c";

export type kitchenSink-list-object-list-select-options = "a" | "b" | "c";

export type kitchenSink-list-object-list-object-select-options = "a" | "b" | "c";

export interface kitchenSink-list-object-list-object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink-list-object-list-object-select-options;
}

export interface kitchenSink-list-object-list {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink-list-object-list-select-options;
  hidden: any;
  object: kitchenSink-list-object-list-object;
}

export interface kitchenSink-list-object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink-list-object-select-options;
  list: kitchenSink-list-object-list[];
}

export interface kitchenSink-list {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: kitchenSink-list-select-options;
  object: kitchenSink-list-object;
}

export interface kitchenSink-typed-list-type-2-object-nested {
  number: string;
}

export type kitchenSink-typed-list-type-2-object-select-options = "a" | "b" | "c";

export interface kitchenSink-typed-list-type-1-object {
  type: "type_1_object";
  string: string;
  boolean: boolean;
  text: string;
}

export interface kitchenSink-typed-list-type-2-object {
  type: "type_2_object";
  number: string;
  nested: kitchenSink-typed-list-type-2-object-nested;
  select: kitchenSink-typed-list-type-2-object-select-options;
  datetime: string;
  markdown: string;
}

export interface kitchenSink-typed-list-type-3-object {
  type: "type_3_object";
  date: string;
  image: string;
  file: string;
}

export interface kitchenSink-typed-list-type-4-object {
  type: "type_4_object";
  name: string;
}

export interface kitchenSink-code {
  code: string;
  lang: string;
}

export interface kitchenSink-code-alt {
  alt: string;
  typescript: string;
}

export interface kitchenSink {
  post: any;
  title: string;
  boolean: boolean;
  map: string;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  color: string;
  colorEditable: string;
  image: string;
  file: string;
  select: kitchenSink-select-options;
  select_multiple: kitchenSink-select-multiple-options[];
  select_numeric: kitchenSink-select-numeric-options;
  hidden: any;
  object: kitchenSink-object;
  list: kitchenSink-list[];
  typed_list: (kitchenSink-typed-list-type-1-object | kitchenSink-typed-list-type-2-object | kitchenSink-typed-list-type-3-object | kitchenSink-typed-list-type-4-object)[];
  code: kitchenSink-code;
  code_alt: kitchenSink-code-alt;
  snippet: string;
}
"
`;

exports[`Output testing should support label, capitalization and delimiter at the same time 1`] = `
"/* eslint-disable */
/* tslint:disable */

export interface Post {
  title: string;
  draft: boolean;
  date: string;
  image?: string;
}

export interface FAQ {
  title: string;
}

export interface SiteSettingsPostSettings {
  front_limit: string;
  author: string;
  thumb?: string;
}

export interface SiteSettings {
  site_title: string;
  posts: SiteSettingsPostSettings;
}

export interface AuthorsAuthor {
  name: string;
  description: string;
}

export interface Authors {
  authors: AuthorsAuthor[];
}

export type KitchenSinkSelectOptions = "a" | "b" | "c";

export type KitchenSinkSelectMultipleOptions = "a" | "b" | "c";

export type KitchenSinkSelectNumericOptions = 1 | 2 | 3;

export type KitchenSinkObjectSelectOptions = "a" | "b" | "c";

export interface KitchenSinkObject {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSinkObjectSelectOptions;
}

export type KitchenSinkListSelectOptions = "a" | "b" | "c";

export type KitchenSinkListObjectSelectOptions = "a" | "b" | "c";

export type KitchenSinkListObjectListSelectOptions = "a" | "b" | "c";

export type KitchenSinkListObjectListObjectSelectOptions = "a" | "b" | "c";

export interface KitchenSinkListObjectListObject {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSinkListObjectListObjectSelectOptions;
}

export interface KitchenSinkListObjectList {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSinkListObjectListSelectOptions;
  hidden: any;
  object: KitchenSinkListObjectListObject;
}

export interface KitchenSinkListObject {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSinkListObjectSelectOptions;
  list: KitchenSinkListObjectList[];
}

export interface KitchenSinkList {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: string;
  select: KitchenSinkListSelectOptions;
  object: KitchenSinkListObject;
}

export interface KitchenSinkTypedListType2ObjectNestedObject {
  number: string;
}

export type KitchenSinkTypedListType2ObjectSelectOptions = "a" | "b" | "c";

export interface KitchenSinkTypedListType1Object {
  type: "type_1_object";
  string: string;
  boolean: boolean;
  text: string;
}

export interface KitchenSinkTypedListType2Object {
  type: "type_2_object";
  number: string;
  nested: KitchenSinkTypedListType2ObjectNestedObject;
  select: KitchenSinkTypedListType2ObjectSelectOptions;
  datetime: string;
  markdown: string;
}

export interface KitchenSinkTypedListType3Object {
  type: "type_3_object";
  date: string;
  image: string;
  file: string;
}

export interface KitchenSinkTypedListType4Object {
  type: "type_4_object";
  name: string;
}

export interface KitchenSinkCodeBlock {
  code: string;
  lang: string;
}

export interface KitchenSinkCodeAltBlock {
  alt: string;
  typescript: string;
}

export interface KitchenSink {
  post: any;
  title: string;
  boolean: boolean;
  map: string;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  color: string;
  colorEditable: string;
  image: string;
  file: string;
  select: KitchenSinkSelectOptions;
  select_multiple: KitchenSinkSelectMultipleOptions[];
  select_numeric: KitchenSinkSelectNumericOptions;
  hidden: any;
  object: KitchenSinkObject;
  list: KitchenSinkList[];
  typed_list: (KitchenSinkTypedListType1Object | KitchenSinkTypedListType2Object | KitchenSinkTypedListType3Object | KitchenSinkTypedListType4Object)[];
  code: KitchenSinkCodeBlock;
  code_alt: KitchenSinkCodeAltBlock;
  snippet: string;
}
"
`;
